### ISSUE TITLE
Differences between QLoRA and this repo

### ISSUE THREAD

**Issue by qwopqwop200** [johnsmith0031/alpaca_lora_4bit#113]
1. Nomal float + Double quantization
QLoRA currently uses zero shot quantization which is different from GPTQ. However, unlike GPTQ, it does not require data, but incurs some performance loss. Therefore, I think the advantage of using GPTQ to train better LoRA is sufficient.
2. Paged Optimizers
Paged Optimizers uses NVIDIA unified memory to avoid the gradient checkpointing memory spikes that occur when processing a mini-batch with a long sequence length. 
3. Apply LoRA to all linear layers
Currently, this repo only applies LoRA to k,v. In case of [QLoRA](https://github.com/artidoro/qlora/blob/main/qlora.py#L290), LoRA is applied to all layers.. This is very important for performance.
<img width="500" alt="image" src="https://github.com/johnsmith0031/alpaca_lora_4bit/assets/64115820/e4052729-0fa7-47d7-8623-e19c1396013f">

4. hyperparameter
The hyperparameters mentioned in the paper are:
"We set LoRA r = 64, α = 16, and add LoRA modules on all linear layers of the base model. We also use Adam beta2 of 0.999, max grad norm of 0.3 and LoRA dropout of 0.1 for models up to 13B and 0.05 for 33B and 65B models."

<img width="488" alt="image" src="https://github.com/johnsmith0031/alpaca_lora_4bit/assets/64115820/4028c92e-8257-4da1-968a-5c514d5d3ee1">

Additionally, 3bit LoRA may be possible. According to the paper "Since finetuning after quantization seems to recover most of the information that is lost during quantization this might enable much more aggressive quantization. For example, 3-bit GPTQ quantization of the basemodel with LoRA might also yield 16-bit full finetuning performance after finetuning"

**Comment by johnsmith0031** [comment]
Great! Maybe we can use larger model at the same performance level of fp16 in the future.

**Comment by Ph0rk0z** [comment]
Qlora perf is terrible. At least 1/2 or 1/3 of the speed of this.

**Comment by gptzerozero** [comment]
What is the advantage of using Qlora? 

I have been using Qlora for finetuning 13b and 30b and wonder if alpaca_lora_4bit can allow me to use a larger context and/or finetune faster.
